### PR TITLE
Yarnberry: parse yarn info output

### PR DESCRIPTION
**Pull request by chmeliik**
STONEBLD-1778

Implement the resolver.resolve_packages function.

Parse the output of `yarn info --all --recursive --cache --json`, pick
out the relevant attributes, post-process them and return them as
Packages.

Alter the Package class slightly:
- make version Optional, use None if yarn info reports 0.0.0-use.local
- make checksum and cache_path both individually Optional, there are
  cases where checksum is null but cache_path isn't
- move checksum and cache_path directly to the Package class
- make locator a string (and rename to raw_locator), because parsed
  Locators don't make a good identifier for error messages (they omit
  unnecessary attributes, so a string representation of a Locator still
  wouldn't be grep-able in the yarn.lock file)
- add parsed_locator property

Remove the unnecessary vet_git_dependencies function, instead validate
that all locators are supported (git locators are not).

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- N/A Docs updated (if applicable)
- N/A Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
Slightly improved the run_cmd error logging and added unit tests for it

**Comment by chmeliik** [comment]
cc @eskultety

We'll need to get you added to the containerbuildsystem org (you can ask @mkosiarc, but not sure if you still want to create that separate GH account)

File: hack/mock-unittest-data/yarn.sh
Comment (chmeliik) [inline hack/mock-unittest-data/yarn.sh:13]: I incorporated the changes from https://github.com/cachito-testing/cachi2-yarn-berry/pull/2

Will update if necessary once (if) that is merged

**Comment by chmeliik** [comment]
Now using more sensible ordering for the keys in the mocked data (the unchanged `yarn info` ordering)

**Comment by eskultety** [comment]
> cc @eskultety
> 
> We'll need to get you added to the containerbuildsystem org (you can ask @mkosiarc, but not sure if you still want to create that separate GH account)

Okay. Nah, tweaked my mail filters for the notifications already instead. Adding me to the org surely will make things easier for me as I'll be able to drop the filters then, but I won't be creating a new GH account for now just for this.

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (lkolacek) [inline tests/unit/package_managers/yarn/test_resolver.py:20]: Do we want to manually fix expected data once they are changed? Would it make sense to make this update part of generate-test-data in Makefile?

**Comment by lkolacek** [comment]
LGTM :+1:

File: tests/unit/package_managers/yarn/test_resolver.py
Comment (chmeliik) [reply tests/unit/package_managers/yarn/test_resolver.py:20]: IMO not really.
- `make generate-test-data` generates the expected output for integration tests.
- `hack/mock-unittest-data/yarn.sh` and `hack/mock-unittest-data/gomod.sh` generate the _input_ for _unit_ tests

There should rarely be a need to update those inputs, but the scripts are there in case you need them

Comment (lkolacek) [reply tests/unit/package_managers/yarn/test_resolver.py:20]: OK agree, we are instructed in a comment that such script to update data is included here, it's enough. :)

**Review by taylormadore — APPROVED** [review]
LGTM :+1:

**Comment by chmeliik** [comment]
Rebased on main, resolved conflicts

**Review by eskultety — COMMENTED** [review]
can't comment on the class attributes and implementation yet, so I have to leave those bits to others.

File: cachi2/core/utils.py
Comment (chmeliik) [reply cachi2/core/utils.py:36]: :+1: fallback on empty stderr sounds better

**Comment by chmeliik** [comment]
Changed the logging of failed subprocesses to only log stdout if stderr is empty

added --depth=1 and --single-branch to the yarn.sh mocking script

File: hack/mock-unittest-data/yarn.sh
Comment (chmeliik) [reply hack/mock-unittest-data/yarn.sh:1]: Good call, it is more readable. I'll fix that in a follow-up PR if you don't mind

Comment (eskultety) [reply hack/mock-unittest-data/yarn.sh:1]: Since this change is related to generating test data sets and hence not crucial to the core logic being added (from functionality POV) I'm totally fine if you submit a follow-up PR for this.
I would have used a different argument if the new file were crucial to the core functionality though :) .

File: hack/mock-unittest-data/yarn.sh
Comment (eskultety) [inline hack/mock-unittest-data/yarn.sh:14]: `--depth` implies `--single-branch` so technically you don't need it. However, explicit is usually better than implicit, so let's keep it this way.

**Comment by eskultety** [comment]
LGTM, still though, someone else needs to grant the actual approval.

**Pull request merged**
